rpc/wrtc_base_channel.go

```diff
@@ -61,101 +61,96 @@ func newBaseChannel(
 
 	var connID string
 	var connIDMu sync.Mutex
-	var peerDoneOnce bool
-	doPeerDone := func() {
-		// Cancel base channel context so streams on the channel will stop trying
-		// to receive messages when the peer is done.
-		ch.cancel()
-
-		if !peerDoneOnce && onPeerDone != nil {
-			peerDoneOnce = true
-			onPeerDone()
+	var peerDoneOnce sync.Once
+	peerConn.OnICEConnectionStateChange(func(connectionState webrtc.ICEConnectionState) {
+		if ch.closed.Load() {
+			return
 		}
-	}
-	connStateChanged := func(connectionState webrtc.ICEConnectionState) {
-		ch.activeBackgroundWorkers.Add(1)
-
-		utils.PanicCapturingGo(func() {
-			defer ch.activeBackgroundWorkers.Done()
 
-			if ch.closed.Load() {
-				doPeerDone()
+		switch connectionState {
+		case webrtc.ICEConnectionStateDisconnected,
+			webrtc.ICEConnectionStateFailed,
+			webrtc.ICEConnectionStateClosed:
+
+			// The Disconnected/Failed/Closed states are all interpreted as terminal. In theory,
+			// WebRTC allows for moving from the disconnected -> connected state if the network
+			// heals or new candidates are presented. We forgo that optimization of preserving an
+			// existing PeerConnection object and instead opt for clients to redial and perform a
+			// new round of signaling/answering.
+			connIDMu.Lock()
+			currConnID := connID
+			connIDMu.Unlock()
+			if currConnID == "" { // make sure we've gathered information before
 				return
 			}
 
-			ch.mu.Lock()
-			defer ch.mu.Unlock()
-			switch connectionState {
-			case webrtc.ICEConnectionStateDisconnected,
-				webrtc.ICEConnectionStateFailed,
-				webrtc.ICEConnectionStateClosed:
-				connIDMu.Lock()
-				currConnID := connID
-				connIDMu.Unlock()
-				if currConnID == "" { // make sure we've gathered information before
-					return
-				}
+			if connectionState == webrtc.ICEConnectionStateDisconnected {
+				// Disconnections happen when the client and server are no longer communicating. But
+				// neither side has closed the peer connection. These events can happen in more
+				// interesting states than typical shutdown (e.g: there may be a bunch of data that
+				// has been queued up to send due to the network problem) such that we feel it's
+				// warranted to log at a higher level.
+				logger.Warnw("connection state changed",
```

Comment:
> Warnw

Why not `Infow`? I worry users will see this log pretty frequently and worry something is wrong if it's at `Warn` level.

Reply:
Warn felt appropriate to me -- nobody asked to be disconnected. And it's exceptional from our perspective where more things can be concurrently happening compared to failed/closed.

But I don't really search by log level -- so warning isn't particularly useful to me. Will move it down to info.